x-pack/plugins/lens/public/editor_frame_service/embeddable/embeddable.tsx

```diff
@@ -159,14 +149,60 @@ export class Embeddable extends AbstractEmbeddable<LensEmbeddableInput, LensEmbe
     render(
       <ExpressionWrapper
         ExpressionRenderer={this.expressionRenderer}
-        expression={this.expression}
-        context={this.currentContext}
+        expression={this.savedVis.expression}
+        variables={this.getVariables()}
+        searchContext={this.getMergedSearchContext()}
+        context={{
+          lastReloadRequestTime: this.externalSearchContext.lastReloadRequestTime,
+        }}
         handleEvent={this.handleEvent}
       />,
       domNode
     );
   }
 
+  private getVariables() {
+    const variables: Record<string, string> = {};
+    this.savedVis.references.forEach(({ id, name }) => {
+      variables[name] = id;
+    });
+    return variables;
+  }
+
+  /**
+   * Combines the embeddable context with the saved object context, and replaces
+   * any references to index patterns
+   */
+  private getMergedSearchContext(): ExecutionContextSearch {
+    const output: ExecutionContextSearch = {
+      timeRange: this.externalSearchContext.timeRange,
+    };
+    if (this.externalSearchContext.query) {
+      output.query = [this.externalSearchContext.query, this.savedVis.state.query];
+    } else {
+      output.query = [this.savedVis.state.query];
+    }
+    if (this.externalSearchContext.filters?.length) {
+      output.filters = [...this.externalSearchContext.filters, ...this.savedVis.state.filters];
+    } else {
+      output.filters = [...this.savedVis.state.filters];
+    }
+
+    output.filters = (output.filters as PersistableFilter[]).map((filter) => {
+      if (filter.meta.indexRefName) {
+        const match = this.savedVis.references.find(
+          ({ name }) => name === filter.meta.indexRefName
+        );
+        if (match) {
+          return { ...filter, meta: { ...filter.meta, index: match.id } };
+        }
+      }
+      return filter;
+    });
```

Comment:
@flash1293 I liked the part where you change the structure of Lens filters to include an `indexRefName`, but I've moved the filter update to the embeddable.